Revist App Integration Tests

SimplyAhmazing opened the pull request:
This PR fixes the broken integration tests and uses lessons learned from [Webtorrent](https://github.com/webtorrent/webtorrent-desktop). Most notably this PR introduces screenshot based integration testing. From the WebTorrent project, the following are listed as benefits of screenshot based testing.

> The integration tests use Spectron and Tape. They click through the app, taking screenshots and comparing each one to a reference. Why screenshots?

> * Ad-hoc checking makes the tests a lot more work to write
> * Even diffing the whole HTML is not as thorough as screenshot diffing. For example, it wouldn't catch an bug where hitting ESC from a video doesn't correctly restore window size.
> * Chrome's own integration tests use screenshot diffing iirc
> * Small UI changes will break a few tests, but the fix is as easy as deleting the offending
screenshots and running the tests, which will recreate them with the new look.
> * The resulting Github PR will then show, pixel by pixel, the exact UI changes that were made! See https://github.com/blog/817-behold-image-view-modes

## Notes

* Use screenshot based testing. DOM checking is used where it makes sense/is easier
* Remove chai and use [tape](https://www.npmjs.com/package/tape)
* Tests for windows and osx. Screenshot based tests for linux have not been passing for me due to inconsistent electron BrowserWindow runing on my local ubuntu VM vs Travis Ubuntu. Linux tests need to be considered elsewhere.
* Tests create a new app data dir called `ZulipTests`. The app normally creates a dir called `Zulip` (`Electron` when started in dev mode) and this usually conflicted with the app integration tests which need a blank slate data directory (i.e. no existing organizations added, clean settings, etc).


## Running the tests

```npm run test-e2e
```
Sample output:

```
Detected App Data Dir base: /Users/moi/Library/Application Support
# app runs
screenshot diff l2 distance: 1887, rms: 19.190186588375322
ok 1 html title
ok 2 screenshot comparison app-basic
Detected App Data Dir base: /Users/moi/Library/Application Support
# add-organization
ok 3 html title
ok 4 screenshot comparison add-organization

1..4
# tests 4
# pass  4

# ok

[15:33:25] Finished 'test-e2e' after 22 s
```

When using screenshots for tests,
* If a screenshot doesn't exist for a test it will be generated and written
* If tests break due to UI changes making a screenshot irrelevant, you can delete old screenshot and re-run test to generate new screenshot
* Sometimes screenshots are not exactly the same, pixel by pixel, and fuzzing algo is used to compare PNG images when by pixel by pixel comparison fails.
* If a test fails it'll also print the `base64` image string so that if the failure is occurring on CI you can convert that string back to the original image. I used [this](http://freeonlinetools24.com/base64-image) an image decoder to debug windows CI screenshots since I don't have a local windows setup.

I added 2 tests -- `app runs` and `add-organization` tests -- hoping other tests will be added.

SimplyAhmazing committed:
use screenshot based testing

SimplyAhmazing committed:
use screenshot based testing

SimplyAhmazing committed:
add zulip opening up screenshot

SimplyAhmazing committed:
add zulip opening up screenshot

SimplyAhmazing committed:
run integration test on travis

SimplyAhmazing committed:
run integration test on travis

SimplyAhmazing committed:
git ignore failed screenshot files

SimplyAhmazing committed:
git ignore failed screenshot files

SimplyAhmazing committed:
revert back to normal window size

SimplyAhmazing committed:
revert back to normal window size

SimplyAhmazing committed:
give zulip test app a separate data dir

When running zulip tests, the appDataDir that electron
creates is normally the same dir as is used in dev mode.

This commit causes the app to create it's own data dir as specified
in config.TEST_APP_PRODUCT_NAME

SimplyAhmazing committed:
give zulip test app a separate data dir

When running zulip tests, the appDataDir that electron
creates is normally the same dir as is used in dev mode.

This commit causes the app to create it's own data dir as specified
in config.TEST_APP_PRODUCT_NAME

SimplyAhmazing committed:
remove hard-coded test app package.json

SimplyAhmazing committed:
remove hard-coded test app package.json

SimplyAhmazing committed:
generate test app package.json file dynamically

SimplyAhmazing committed:
generate test app package.json file dynamically

SimplyAhmazing committed:
remove unneccessary console log

SimplyAhmazing committed:
remove unneccessary console log

SimplyAhmazing committed:
remove cursor before taking a screenshot; remove cursor from saved screenshot

SimplyAhmazing committed:
remove cursor before taking a screenshot; remove cursor from saved screenshot

SimplyAhmazing committed:
add screenshot for linux tests

SimplyAhmazing committed:
add screenshot for linux tests

SimplyAhmazing committed:
log W x H dimension comparison failure

SimplyAhmazing committed:
log W x H dimension comparison failure

SimplyAhmazing committed:
log base64 image string

SimplyAhmazing committed:
log base64 image string

SimplyAhmazing committed:
log b64 img string w/o reading file

SimplyAhmazing committed:
log b64 img string w/o reading file

SimplyAhmazing committed:
log screen resolution on travis

SimplyAhmazing committed:
log screen resolution on travis

SimplyAhmazing committed:
add xrandr extensions to Xvfb

SimplyAhmazing committed:
add xrandr extensions to Xvfb

SimplyAhmazing committed:
get travis screen dim with xdpyinfo

SimplyAhmazing committed:
get travis screen dim with xdpyinfo

SimplyAhmazing committed:
start xvfb using start-stop-daemon in before install on travis

SimplyAhmazing committed:
start xvfb using start-stop-daemon in before install on travis

SimplyAhmazing committed:
remove attempt to install xrandr extension

SimplyAhmazing committed:
remove attempt to install xrandr extension

SimplyAhmazing committed:
remove exit call from travis build script

SimplyAhmazing committed:
remove exit call from travis build script

SimplyAhmazing committed:
call xvfb init after start-stop-daemon on travis

SimplyAhmazing committed:
call xvfb init after start-stop-daemon on travis

SimplyAhmazing committed:
cut app height in half

SimplyAhmazing committed:
cut app height in half

SimplyAhmazing committed:
reduce app min height

SimplyAhmazing committed:
reduce app min height

SimplyAhmazing committed:
update linux screenshot to img w/height=300

SimplyAhmazing committed:
update linux screenshot to img w/height=300

SimplyAhmazing committed:
verify height offest on travis

SimplyAhmazing committed:
verify height offest on travis

SimplyAhmazing committed:
temporarily comment out electron-window-state pkg

SimplyAhmazing committed:
temporarily comment out electron-window-state pkg

SimplyAhmazing committed:
update linux screenshot

SimplyAhmazing committed:
update linux screenshot

SimplyAhmazing committed:
setup windows build

SimplyAhmazing committed:
setup windows build

SimplyAhmazing committed:
add dir for win screenshot

SimplyAhmazing committed:
add dir for win screenshot

SimplyAhmazing committed:
go back to using window state manager

SimplyAhmazing committed:
go back to using window state manager

SimplyAhmazing committed:
Merge branch 'revisit-app-integration-tests' of github.com:zulip/zulip-electron into revisit-app-integration-tests

SimplyAhmazing committed:
Commit to trigger Appveyor

SimplyAhmazing committed:
use appveyor.yml file from master

SimplyAhmazing committed:
commit to trigger appveyor

SimplyAhmazing committed:
triggr appveyor build

SimplyAhmazing committed:
trigger appveyor build

SimplyAhmazing committed:
removes foo.txt

SimplyAhmazing committed:
setup app data dir for windows

SimplyAhmazing committed:
log windows image as b64 image

SimplyAhmazing committed:
add image for test

SimplyAhmazing committed:
remove smaller sized screenshot

SimplyAhmazing committed:
comment out window state keeper; add screenshot

SimplyAhmazing committed:
add windows basic screenshot

SimplyAhmazing committed:
bring back window state keeper

SimplyAhmazing committed:
comment out unused test func; add comment to func

SimplyAhmazing committed:
remove chai/chai-as-promised deps

SimplyAhmazing committed:
fix formatting

SimplyAhmazing committed:
fix comment

SimplyAhmazing committed:
clean tests package

SimplyAhmazing committed:
add test adding an organization

SimplyAhmazing committed:
find tests with gulp

SimplyAhmazing committed:
add screenshot for add organization test

SimplyAhmazing committed:
remove scrollbars from webviews

SimplyAhmazing committed:
reset test data dir at the start of a test only

SimplyAhmazing committed:
update add organization test screenshot to screenshot w/o scrollbars in webviews

SimplyAhmazing committed:
remove unused TEST_DIR config

SimplyAhmazing committed:
fix indentation

SimplyAhmazing committed:
add gulp cmd test-e2e; dont run e2e tests on linux

SimplyAhmazing committed:
PR cleanup

SimplyAhmazing committed:
remove tests package.json

SimplyAhmazing committed:
add comment

SimplyAhmazing commented:
Link to build passing on AppVeyor: https://ci.appveyor.com/project/SimplyAhmazing/zulip-electron/build/build29

akashnimare commented:
on my macOS, `add-organization` tests fails.
<details>

```sh
npm run test-e2e

> zulip@1.5.0 test-e2e /Users/akka/dev/zulipwork/zulip-electron
> gulp test-e2e

[02:08:05] Using gulpfile ~/dev/zulipwork/zulip-electron/gulpfile.js
[02:08:05] Starting 'test-e2e'...

TAP version 13
Detected App Data Dir base: /Users/akka/Library/Application Support
# app runs
screenshot diff l2 distance: 2491, rms: 23.649098186423174
ok 1 html title
ok 2 screenshot comparison app-basic
Detected App Data Dir base: /Users/akka/Library/Application Support
# add-organization
screenshot diff l2 distance: 52583, rms: 71.04397999165795
ok 3 html title
not ok 4 screenshot comparison add-organization
  ---
    operator: ok
    expected: true
    actual:   false
    at: process._tickDomainCallback (internal/process/next_tick.js:129:7)
Saving screenshot, failed comparison: /Users/akka/dev/zulipwork/zulip-electron/tests/screenshots/darwin/add-organization-failed.png
Failed Image: iVBORw0KGgoAAAANSUhEUgAAA+gAAAJCCAYAAABTU+3YAAAgAElEQVR4nO3dWYxd94Hn99+599bKrcjiUlzEnaIWaqFla7PaarftUXrJdDoBgp4gCxIkCAIkCDIPAwR5y0OQtyCTlwHmJQECpIHOZNKDeHrctmxL3bJaEmVK1mZKJEVREvetirVX3XvyQEq2xEUkq8r1L+rzAa6We889519P5Lf+//M/1c59++vcJQ68+PxiDwEAAADuSGOxBwAAAAAIdAAAACiCQAcAAIACCHQAAAAogEAHAACAAgh0AAAAKIBABwAAgAIIdAAAACiAQAcAAIACCHQAAAAogEAHAACAAgh0AAAAKIBABwAAgAIIdAAAACiAQAcAAIACCHQAAAAogEAHAACAAgh0AAAAKIBABwAAgAIIdAAAACiAQAcAAIACCHQAAAAogEAHAACAAgh0AAAAKIBABwAAgAIIdAAAACiAQAcAAIACCHQAAAAogEAHAACAAgh0AAAAKIBABwAAgAIIdAAAACiAQAcAAIACCHQAAAAogEAHAACAAgh0AAAAKIBABwAAgAIIdAAAACiAQAcAAIACtP7gz59b7DEAAADA154ZdAAAACiAQAcAAIACCHQAAAAogEAHAACAArQW4qT9y5ZlcN26rB8ayurBwfT196eqqkxOTuTCufM5c+pkzp05k/HRsYW4PAAAACw58xboVVVl3dBQ9tx3X/r6+9PV1ZVms5mq0UhVVamqKn39/dm0pTcbNm3M7MxsJsbH8/677+b0iROp63q+hgIAAABLzrwE+opVq3L/g/syMDiYru6uNKoquRrlSVJdPa6qqlTNZhrNZrq6utPb15fHnnoyF89fyNsHD2bk0qX5GA4AAAAsOXMO9KFNm/PQ/v3p7u5O1bi9W9qrqkp3d082bNqY1YNrcvDVV3Pi+MdzHRIAAAAsOXPaJG7rjp3Z/63H09PT8/ls+Z2oqio9vb15/JlnsuPePXMZEgAAACxJdxzoQ5s25YGHHkqj8dVhfqt3lzcajTzyzW9m09Z77nRYAAAAsCTdUaCvWLky+x55NI0vLWmv6zp16nTqz16ddOpO6quvTuc3r7rTSa6zMVyj0cj+xx/PyoGBO/uJAAAAYAm67XvQq6rK3gceTFd39xfer1PnxeMH89bJQ5mZmEim28lMO5meTT3TTj199TU7m85sJ8u2bcg//M6fZNe6rfnNNnJXdHV3Z9/+R/Pyz1+wuzsAAABfC7cd6GvXb8iq1Wuuuee8SpX1/aszOTOZU5dOJVOzV18z6UzNpp6cTWd6Npltpx5clu705L3zx7Jh1bqs6F32hXM1Go2sXrs2GzZtyqlPP53bTwgAAABLwG0vcd+15950dXVd97PNK9bn3jXb0vzS0vfq6j+qZpVqoD/Njasy22jn3bNHc3L03HXP1d3dnb37Hrzd4QEAAMCSdFsz6H39/enr78uNNmxf0bMs96/flU8ufJrpiamry9xnU0/NZnxsNMNTo2mv7k0aVep2nZHxkXx44ZPcu3bbNeeqqir9y5alf/nyjI+O3tEPBwAAAEvFbQX6msG1aTVb+fI945+pqip71m3PnvXbklSfL4OfmpnKG0d/lb/+1c8yMj2WzNZpNpKhvtXZMbDpxoNrtbJ2/focF+gAAADc5W5rifu69evTaDa/8rjqapxXVZV2p50Dh17Pj1//aS6PjiTtTup2O9tXbcxz9z6TXYM3fqRas9nMhk0bb2eIAAAAsCTd1gz6wOrV12wOdzOzs7N59d1X88LBFzM8Ppy6UaeRrmxZuyl/uv+5bFw9lEZ14+CvGo2sWbf2doYIAAAAS9JtBXpPb98tBXqdOtPT0znw1mv5+cEXcnHsUtpVne7engwNbMgf7f9+hgY2pKfVfdPzVFWVvr7+2xkiAAAALEm3/Zi1r1KnzsT4WN5+/+385Bd/k5GJkatx3psNK9fmB9/4g+zZtDOt5vV3ggcAAICvo9sK9KnJifT29d9wFr2u64xPjOXwh+/nhz/5qwyPj6RuVOnq6c7alYN57ol/kId27rvlZfJ1XWdiYvx2hggAAABL0m1tEnfp4sXUdX3Dz6dnpvPOe7/K//2v/iLnL5xLZ3Y2VaeT1ctX5buP/X4e2PlA6tTp1PVNz/OZutPJhbPXf046AAAA3E1uawb97JkzWbd+w3V3cq/rOr9867X8m5/9MJdGLyWNKkmVvt7+PPHIU9kydE8uDF9IVSepk2arleXLVqSnu+eG12u32zl94uTt/kwAAACw5NxWoF84fy6z7dm00pUvPwu9qqqMdbWTLavStSpJu5N0Opmtq7z41kt55b3Xkk4nma1Ttzvpqpr50+f+vey7/+FUuf5O7rOzszl35syd/mwAAACwZNxWoE+Mj2difOLqbu7Xfr574478+tQHGZ0eS2Y7yWw7ndlOLk+MZmS2feUZ6DNXnoPe6CTvHHor92zemtWrB685V13XGR8by/jo6B3/cAAAALBU3NY96Ely5IP3MzMzc93P1q9cm/s27kl3qztVrsyqV6lSVVUaVePK/1dVGlWVuq7z/tFf5+MTx697runp6Rx6+53bHR4AAAAsSbf9mLVzZ05n+OKFrNuw4Zrd2Pt7+rNnaGeOnPow4xNjV5a5z3ZSd+or/93upL66xD3tTrqaXTlz9nTquv7CuTqdTi6eO5fTJ07M/ScEAACAJeC2A72u6xx6950MrF6d7p5rN3jbtu6e/Jc/+E9SV/k8uqskVaNKqir5fPP26ur7jWtCf2Z6Om8ffOOWdnoHAACAu8FtL3FPkssjI3n7zTfS6XSu+/lnS9o/e1VVI1Wu/n/jN6/rxXmn08nBV1/NyKVLdzI0AAAAWJLuKNCT5NSJE3n3rbfS6Xz1LPd19pO7rk6nkzcPHMiJ4x/f6bAAAABgSbrtJe6/7fiHRzM9NZWH9u9Pd3f3NbPht6qu60xPTeXgq6+KcwAAAL6W5hToSXLqxKcZGxvN/Q/uy8DgYLq6u67eXf7V6rrOzMx0Lp6/kLcPHrSsHQAAgK+tOQd6klweHs5rL/8i64aGsue++9LX35+urq40m81UjUZSXUn2uq5TdzppdzqZnZnNxPh43n/33Zw+ccKGcAAAAHytzUugJ1fi+8zJkzlz8mT6ly3L4Lp1WT80lNWDg+nr70+STE5M5MK58zlz6mTOnTmT8dGx+bo8AAAALGnzFui/bXxsLONjY/n42LGFOD0AAADcde54F3cAAABg/gh0AAAAKIBABwAAgAJUO/ftv2u2Tz/w4vOLPQQAAAC4I2bQAQAAoAACHQAAAAog0AEAAKAA8xLovcv60ur64iPVm61m+pb3zcfpAQAA4K7X+upDbu7RZx/Lmo1rkyTvvPyrnP7oZAY3rs0jzz6WJBk5P5zXf/JK6vrGe9E9+PTDqaoqb7/05lyHAwAAAEvSnGbQu3q6s3poMC/91Qv59WvvZNfDe5Ik2x/clWPvHMmL/+L5rFizMr3Leq/7/Z6+3uz/7reyYevG5K7ZSx4AAABu35xm0Nuz7dSdOvc/uS8zUzPp6e/N7kf3ZuWaVZmZms79Tzz0+XFfVlVVHv+3ns7k+ETOnzyXRtPt8AAAAHx9zamKO+12Xn/+laROVqxekaqqMrBudapGlRVrVqXV1crBnx3I9OT0Nd+t6zoH/ublvPajlzNxeSxJNZehAAAAwJI252nryxdG8sbPD+S9V95Op93JgR//fSbHJnLowLs5+LPXcunMhRt+d2JsIsmV2XQAAAD4OpvzJnEbd2zKfd/al6pxJbL/4M+fS5I8/Hv7U9d1jrz5QY7/+sO5XgYAAADuanOaQW91tXLft/bl2LtHc/iNQ6k7dd584fW0Z9s5/usP88Evf51dj+xJT1/PfI0XAAAA7kpzCvSq0Uiq5NSxT5MkE2PjOX/yXEYuDCepcvr4qSsXadgADgAAAG5mTkvcZ6amc/H0hTz1J99Jkrz3yttJkuPvfZhHnn0sW+/bntFLlzM5PnnT83jCGgAAAF931c59++fcx33L+9KebX9ht/aunq60uroyMTo+19PfsgMvPv87uxYAAADMpzlvEpckE6MT17w3MzWTmamZ+Tg9AAAA3PXcHA4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQgNZCX2DZsv70dPdkanoqY2PjC305AAAAWJIWPNC///vP5sH778s77/06f/XDv17oywEAAMCSZIk7AAAAFECgAwAAQAEEOgAAABRAoAMAAEABBDoAAAAUYF52cf/sUWrX09PT8/m/16xefd1jPIINAACAr7t5CfTPHqV2M7t37sjunTuu+5lHsAEAAPB1Z4k7AAAAFGBeZtB/8vMX8re/+Pvrfvb97z6b3Tt35PDRD/OTn71w3WOmpqfmYxgAAACwZM1LoI+Njd/wHvKpqanP/33h4sX5uBwAAADcdSxxBwAAgAIIdAAAACiAQAcAAIACCHQAAAAogEAHAACAAszLLu4389kj2DxKDQAAAG5swQP9Zo9gAwAAAK6wxB0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAArcUeAABwZ+ok7brOx5dGkiRbB1amUVWpFndYAMAdMoMOAAWok3Tq+vNXXdepv+o7dZ2Tl8fyv7zwcv7p376SM+OTv4uhAgALxAw6ACyyTpLhyan86tTZHDl3IRtWLMujGzdksL8vva3mTb87MjmVd0+dSavRyOXJqQwt6/vdDBoAmHcCHQAWUbtT5/zEZP7PX76V/+vAGxmemEx/V1ee2bM9/+R738mW5cvSuMGa9SpVtq9elf/48W+k0aiyedWKW7pmp64zOdtOu9NJb1crzapKo7IwHgAWm0AHgEU03W7nhaMf5S8OvJFPLg6nrpPLk1P5lwffyY7BNflvvv2ttHL9gK6qpL+rlX/0yANJklaz8ZX3n9d1Mj4zm58f/TjHLl7Md3Zuy72Dq79yph4AWHgCHQAW0UynnSPnLuTc6Hjanc6VN+ukneTgxycy1e6kdZN4rpJ0NW99S5k6dT4ZHsn/8crr+eDMuZwcHsk/fvYpgQ4ABRDoALCoqjSbzc//O6lTpUpVVWk1m6nmeel5naSr0Uyj0chMu53mAlwDALgzAh0AFlFvq5Untm7Oj9evza8+PZG6U6XZrLKytzfP3b87fc3GvAZ0o6qyfXAg//O//YN8Mnw5j23ZmC59DgBFEOgAsIiajSoPD63Lf/HUY/l/ftWfc6NjWdnbk6d2bM0f7Nox7881r5I0k2xdtSJbV61I0+w5ABRDoAPAImpWVQZ6e/IPH7w3T2zdnPPjE1nW3ZXNq1akr9Wa1zj/8nUBgLIIdAAoQHejkc0rV2TjiuWpqisB7d5wAPh6EegAsIA6dZ2ZTiftTp3kyo7r13vueFVVaVZJc8HmzAGA0gl0AFhAYzOzef3TU/nwwqWkTvZv3pC96wbT57FmAMCXCHQAWAB1rsT5Xx86kn/+0qs5fmE4ders37Ix//nT38p3d25LszG/G8DNp05dZ6LdydsnTyepsnPNqqzu602rcevPXAcAbo9AB4AFUNd1PrownP/95QN55+TpzLY7qVLl744ez+pl/dm/eShrenuKvc+8rhp54eiH+R//9U/T19XKf/r0Y/mzB/dmRXf3Yg8NAO5afg0OAAvk1OXL+XT4cuort5+nTp3ZdjunR0YzPDH5+fslGpuezvOHjuT0yEg+vjScl44cz+Rse7GHBQB3NYEOAAtkcFl/1i5flsbVSfKqqtLdbGbNsv70d3Wl2PXtSVqNRh7dsjFDq1Zl06oVeWBofbosbweABWWJOwAsgKqqsnPNQP6dhx/I/3bxUoYnJtOoqmwaWJkf7N2dtcv6i34WeV+rmT99cG+2rFyZZrPK3rWDWdFjeTsALCSBDgALoErS39XKf/CNh7JtcCCvHfskrWYzz92/J3vXrk6rUW6cJ1fGv7yrlWe2b0k8lx0Afieqnfv2F3wH3O058OLziz0EAPiCOld2RK/rOnWdNBvXPgMdACAxgw4AC6pKfhPkVcxCAwA3JNABYIFVSSLMAYCvYDtWAAAAKIBABwAAgAJY4g4AC+C3d2C1uB0AuBVm0AFgHlVVlarRzHi7k3OT0+l4PBkAcIvMoAPAPDo7Np6fHj6Wf/nmuxmdms7mgVX5z578Rh4eWpfeVnOxh3fH2p06o9MzGZ6eSl+zlRU9Xelt+WsEAMwnf7ICwDx69eMT+ac//0XeP3M2SZX+7q5Mt9v577//TPYMrl7s4d2ROsnIzHT+17/9+7x4+FgeGNqQ//r3Hs99a9cs9tAA4K4i0AFgHv39R5/k9OXRXLnzvMpMu5NDp8/k2MXh7B5cvSTvR2/Xdd48cSb/5p33c2L4cj65OJxHN2/ItoGV6TOLDgDzxj3oADCPulutL9xzXiVpNhppNpbwH7l10mo00mo002w00mo20tVsLslfNgBAyZbw3xYAoDzP7NyaLatXXd26vU53q5WHt2zMrsGBJRu0zUaVh4fW5z984hv55rbN+fNvPprv7tmZ3lbXYg8NAO4q1c59++uvPmxpOPDi84s9BAC+5i5NTuWNk6fzlwffybmxsdw3tD7/aP9D2bF6ZXqaS3iTuLrO2NR0ZjudNBuN9HW10r2Efx4AKJFAB4B51mg00mw2U1dV6k6dutNOp9NZ7GEBAIWzswsAzLNOp5NOffX33/Vd83twAGCBCXQAWAjCHAC4TTaJAwAAgAIIdAAAACiAQAeA34H66gsA4Ebcgw4AC6xOUn92T3pVLdnnoQMAC0ugA8ACqZOMz8zm8PmLeffMuTSrRp7avjmDvT3p7+pa7OHdVF0n0+12Lk1Np9mo0tdqprfVSrPy6wUAWCgCHQAWSKdODp48nf/hX/04H5w9lyrJM7t35J9875l8Y9OGomfSZ+s6B0+ezl8efCfdrWa+f++uPL1tc/q6/NUBABaKP2UBYAHUdTI6M5N/8cY7OXb+QtqdTqokb35yIj99/0geXL82Pa1mkZFeJzk1Op5/9nev5aWjH6VKcvjs+exe+1y2DqwscswAcDcQ6ACwIOrMtDs5fXk07U4ndZ1UVTI9O5sL4xOZrev0LPYQb2J0ejqnRi5nanY2dV3nxPBIxmdnF3tYAHBXs4s7ACyEqsqyrla+tfWedLdaqao6VVVlVX9/7htan/6uVrEz0VWSzSuX58md27Kitzurl/XnqZ3bs3758lTuQQeABWMGHQAWQJWk1ajyZw/fl5HJibx78nSSRn5v97Y8t3dXsXH+mf6uVv6rp7+ZB9YPpqqq7N+yMSu6mlfW7gMAC6LauW//XfMn7YEXn1/sIQDAF7Q7dS5NTuXU6GhSJ9vXDGRZd5fQBQCuYQYdABZQo1FldV9PVvV2J8mVx5SJcwDgOgQ6ACygKklVVWm4dxsA+Ao2iQOAUlRVptqdzP4OZtjr33oBAGUwgw4Ai6xOcnZ8Ij87fCwfnDmXoZUr8v17d2bTimXpbjbn9Vqdus5Mp85Hl0ZydmwsOwfXZE1vT3qafmcPAItNoAPAIqrrOpPtTv7ft36df/6LV3NpfCr93V358MLF/Le/92TWL+ub1+tNtdt55fiJ/LOXXsupkcvZvW4w//i7T+f+dYPF7ywPAHc7vy4HgEXUrut8Onw5Lx35KJ9eHM7lyamcvTyWv3j9rXx6eXTer3dxYip/+cY7OfDRJzl85lx+euhI/vW7H2S2ttwdABabGXQAWERVqjQbVRqNpE6Vuu4kjUZ6Ws20vmLZebtTZ6auMzI5lVTJqt7etKqrO8XfQKPRSKeu06nrtOtOOnUns51O6rpOZSM7AFhUAh0AFlFVJeuW9+fxbffklWOfZHRqOv3dXXnugXuzpqfnpt+dbLfzo/eP5qWjH6VKld+/d2e+t2tb+lo3vm+9v6uZ7+zekbdOnMqJS8PZuW5tnti2JV0NcQ4Ai02gA8AialRV+lut/PuP7suu9Wvzxicnc8/Aqnx7xz1Ze5P7z6uqymsff5r/6Ucv5OL4WJI6f3fkWNb+u3+UJ7duuuH3+ltd+eP7d+e+oXW5MD6eLatWZevK5e4/B4ACCHQAWGSNqsrq3u58d8c9eWbr5lRV0tts5WYrzpvNZn599nxODA+n3WmnSpXJ2eEcOns+T2/fkk6nc93vtRpVVnZ35eH1g0kG06gqcQ4AhRDoAFCARlWlUVVpNa7cd/5V0dzpdLJ55cos7+nJyMREqqrKqt7eDK1YfsM4/203u08dAFgcAh0ACnKr2dzpdPLkts3577777Tz//pE0qirP3b8739yycUHHBwAsHIEOAEvUmr7e/EePPZTv79meKlU2rVqR3ptsEAcAlE2gA8AS1ayqLO/uyp61axZ7KADAPLj5A1YBAACA3wmBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABRDoAAAAUACBDgAAAAUQ6AAAAFAAgQ4AAAAFEOgAAABQAIEOAAAABVjQQK+qKstXrsr6zfdk+cqBVFV1w2NbXd0LORQAAAAoWmshTrp5+65s3XNfVg6sSbP1m71dVagAAAhQSURBVEu027O5fOliPj7yfj4+8v7n7/ctW56HHv92Xv3ZjxZiOAAAAFC8eQ307p7e7Hv86Qxt2Xbdz5vNVgYG12VgcF2G7tmWt155KVWjkSe/94fpdDrzORQAAABYUuYt0JevGsiT3/vDdPf0JknOnTqREx8dzcjF8xm7fDnLV6zMyjWD2bx9V9asH8q6jVvye3/0Z2nPzqS3f1nGLo/M11AAAABgyZmXQK+qRh556jvp7ulNe3Ym7x18LccPH/rCMcMXz2f44vl8fOT97Nj7YPY+8li6urvT1e3ecwAAAJiXTeJ2P/hwVq0eTJIceOEn18T5l5365KNMTU3Ox6UBAADgrjDnQO/tX5bd+x5Jkhw79G7Onzl10+P7li3Pk9/7w/T1L5vrpQEAAOCuMedAHxhcm6pqpN1u59CbB2567Odxvmz5XC8LAAAAd5U534O+8urS9suXLqTdbt/4Ql3deejxb6fT6Vx3Q7iJsctzHQoAAAAsWXMP9IE1SZKRixduetzszLTnnAMAAMANzMsmcQAAAMDczDnQRy5dmTlfuXrNnAcDAAAAX1dzXuI+cvF8kmTFwJo0m82b3of+Zc1WV/Z966lUVZXjhw/lwlfsAA8AAAB3qznPoF86fy513Umz2czeR755W9/dsfeBbN6+Kxu2bM3lSze/hx0AAADuZnMO9MnxsRx++80kyfa9D2Rw/dAtfW9gcF123v9QkuTTD49kZnp6rkMBAACAJWteNok7/M6vMnx1qfs3n/1+tu7ee9Pj1w5tyuPffS6trq5MjI3m8NtvzMcwAAAAYMma8z3oSVLXnbz58ot58nt/mO6e3uz71tMZumd7Tnx0NCMXz2dsZCS9/f1Zvmog2+99IIMbNiZJpicn8+rPfpTJifH5GAYAAAAsWdXOffvr+TpZd09v9j3+dIa2bPvKYy+dP5u3Xn0ply9dnK/L58CLz8/buQAAAOB3aV5m0D8zPTWZX/7tT7N5+65s3XNfVg6sSbP1m0vUdZ3RkUs5/PabOXn8w/m8NAAAACxp8xron/n02JF8euxIqqrKshUrs2zFqkyMj2ZsZPi2HsMGAAAAXxcLEuifuTJjPpzRkeGFvAwAAAAsefOyizsAAAAwNwIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAAIdAAAACiDQAQAAoAACHQAAAAog0AEAAKAAAh0AAAAKINABAACgAK2FOnH/8hXZ89D+VFX1+Xt1XeeDtw5mfPTyQl0WAAAAlqQFC/TN23dl8/Zd17w/dnkkh99+Y6EuCwAAAEvSgi1x71+x8rrvL7vB+wAAAPB1Ni+B3mhce5obhfj13r/e9wEAAODrZM5L3B96/NvZtG1nTh7/MMePHMrM1FTu/8bjGRhcd93jBwbX5ZvPfj/v/fLVdPX0ZOuuvdm4dUdOfHQ0b7360lyHAwAAAEvSnAJ9YHBd7tl1b5Jky8492bJzzy19b/2me7J+0z1feO+eXffm4yPv59L5s3MZEgAAACxJc1pb/sBjT8zXOBbkfAAAALBU3HGgb9m554bL2L9sdmbmlo4bGFx3y7PwAAAAcDe54yXuQ/dsu+nndaeTD955Mx8fPpSpyYl09/Zm6+77sufBR1LdZFO4oXu25ZOjH9zpsAAAAGBJuuMZ9HdffyXDF8/f8PM3Xn4hh99+I1OTE0mS6cnJHH77jbzx8gs3/M7wxfN59/VX7nRIAAAAsGTdcaCPj17Oyz/+YT798PA1n507dSInjx+77vdOHj+Wc6dOXPP+px8ezss//mHGRy/f6ZAAAABgyZrTJnGddjvHDx+65v2LZ0/f9HvX+/z44UPptNtzGQ4AAAAsWXMKdAAAAGB+zCnQG81mtu7ee837q9dtuOn3rvf51t1702g25zIcAAAAWLLuOND7l6/IUz/442zesfuaz9YObcrGrduv+72NW7dn7dCma97fvGN3nvrBH6d/+Yo7HRIAAAAsWXcc6A889kRWrR684eePPvVsdu97ND29fUmS7t7e7N73aB596tkbfmfV6sE88NgTdzokAAAAWLLu+Dnopz7+KOs33XPDz6tGI/c+tD/3PrQ/szMzaXV13fJ5AQAA4OvmjmfQPzn6QS6dP3tLx95qnF86fzafHP3gTocEAAAAS9acNol79/VX5mscC3I+AAAAWCrueIl7cmXG++Mj72fTtp05efzDHD9yKDNTU7n/G4/fdPn7mRMf571fvpqunp5s3bU3G7fuyImPjt7yjDwAAADcbaqd+/bXcz1Jo9FIp9P5wntP/4M/ycDgumuOvXT+bH7xN//fV37/Thx48fk5nwMAAAAWw5yWuH/menE9dnnkusde7/35iHMAAABYyuYl0K9n/DYCHQAAAL7u5nQP+s18euxI+lesTFVVn79X13VOHDuyUJcEAACAJWvBAn189HLefPnFhTo9AAAA3FUWbIk7AAAAcOsEOgAAABRAoAMAAEABBDoAAAAUQKADAABAAQQ6AAAAFECgAwAAQAEEOgAAABRAoAMAAEABBDoAAAAUQKADAABAAQQ6AAAAFECgAwAAQAEEOgAAABRAoAMAAEABBDoAAAAUQKADAABAAQQ6AAAAFECgAwAAQAEEOgAAABRAoAMAAEABBDoAAAAUQKADAABAAQQ6AAAAFECgAwAAQAEEOgAAABRAoAMAAEABBDoAAAAUQKADAABAAQQ6AAAAFECgAwAAQAH+f1qdl7ABvPjpAAAAAElFTkSuQmCC
  ...

1..4
# tests 4
# pass  3
# fail  1

[02:08:23] Finished 'test-e2e' after 19 s

npm ERR! Darwin 16.7.0
npm ERR! argv "/Users/akka/.nvm/versions/node/v6.9.4/bin/node" "/Users/akka/.nvm/versions/node/v6.9.4/bin/npm" "run" "test-e2e"
npm ERR! node v6.9.4
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE
npm ERR! zulip@1.5.0 test-e2e: `gulp test-e2e`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the zulip@1.5.0 test-e2e script 'gulp test-e2e'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the zulip package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     gulp test-e2e
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs zulip
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls zulip
npm ERR! There is likely additional logging output above.

```

</details>

timabbott commented:
BTW, huge thanks for working on this @SimplyAhmazing!

SimplyAhmazing committed:
only start xvfb on linux ci

SimplyAhmazing committed:
start xvfb from script on ci

SimplyAhmazing committed:
remove .python-version

SimplyAhmazing committed:
use looks same; modify image diff logging

Update tests to use looksSame npm pkg instead of custom algorithm
for comparing test screenshots.

Log base64 image diffs only when running on CI. Otherwise save image
diff locally into *-failed.png

SimplyAhmazing committed:
Update test screenshots

Remove old screenshots because they blurred focus from input elements.
The current screenshots contain the caret (blinking cursor) on input
elements and rely on looksSame() to filter the caret out.

SimplyAhmazing committed:
ignore test generated package.json file

SimplyAhmazing committed:
update mac test screenshot (page scrolled to lower point)

SimplyAhmazing committed:
update screenshot for add-organization

SimplyAhmazing committed:
update to travis screenshot

SimplyAhmazing committed:
ensure window exec JS promise is resolved

SimplyAhmazing committed:
update mac screenshot; increase timeout time for test

SimplyAhmazing commented:
Made some updates to this PR after following up on the error Akash got in running the tests locally. I have to come to learn a lot about Visual Regression Testing that I want to share here and discuss with you all. The things I want to share are organized in these topics,

* Non deterministic rendering of images
* Electron/Chrome non deterministic states
* Visual regression work flows
* Tools & Literature


## On Non deterministic rendering

I was aware that rendering is non-deterministic across platforms (win vs mac vs linux). It turns out rendering is inconsistent on the same platform (e.g. my Mac book air vs Travis OS X build worker). These differences  can occur in several layers in the stack (hardware, OS, Application GUI framework). Specific things include GPUs can generating different pixel values due to floating point calculations, display DPI, how anti-aliasing is handled, how fonts are rendered, etc.

From [reading](http://twolfson.com/2014-02-25-visual-regression-testing-in-travis-ci) around I’ve learned that the best way to control for these sorts of issues is to use near identical environments for generating screenshots and comparisons.

Before I was using a custom algorithm for comparing screenshots (borrowed from WebTorrent) but I found the library [looks-same](https://github.com/gemini-testing/looks-same) is much better and can accounts for,

* Anti-aliasing differences
* Ignoring the caret on input fields (the blinking text cursor)
* DPI adjustments

`looks-same` can also generate screenshot diffs to show you where the comparison failed which was really helpful in debugging.

There’s another popular JS library for generating diffs, [Resemble.JS](https://github.com/Huddle/Resemble.js) but it required `Node 8`. It may perhaps have better anti-aliasing support than `looks-same` but I haven’t been able to compare thoroughly.

## Electron/Chrome non-deterministic rendering states

Sometimes the Zulip app doesn’t render consistently. Two things I came across are,

* scroll-bar auto-hiding - When the automated test added a server URL and the page navigated to the login page for a community. Sometimes the scroll bars in that web view would be hidden and sometime they would be displayed. In general this is a feature of Chrome where it can auto-hide the scroll bar for you. I wasn’t able to find a setting where you can force electron be consistent with this setting. I was able to get around it by injecting CSS to auto-hide the scrollbars on the page before every screenshot.
* Auto-focus location - When you go to the login page, the login page will not be scrolled to the same exact location. I was talking to @Lev about this and says it seems that [Zerver auto-focuses on the input element](https://github.com/zulip/zulip/blob/9cc561e442fe441ba8a775e444a1399010cc03ab/templates/zerver/register.html#L175). I noticed that the scroll location can be off by a pixel or two upwards or downwards between test runs. I was able to get around this by telling the webview to the bottom before a screenshot is taken.

There may be more such intricacies we’ll have to handle as they come up.

## Visual Regression Testing Workflows

My research on visual regression workflows shows that they almost always require human intervention to overcome technical issues in image comparisons. Things like font rendering, blinking cursors, scrollbars / scroll location, anti-aliasing in sensitive areas such as a logos/icons don’t equate to a reasonable test failures. For these sort of differences a human operator can choose to ignore the subtleties reported by the image comparisons.

[Brett Slatkin](https://github.com/bslatkin/dpxdt#how-to-use-depicted-effectively) has this simple workflow that they used at Google which I’ve copied here:

1. Establish a baseline release with an initial set of screenshots of your site.
2. Create a new release with a new set of screenshots of your new version.
3. Manually approve or reject each difference the tool finds.
4. Manually mark the new release as good or bad.
5. Repeat. Your approved release will become the baseline for the next one.

There are tools that are used to facilitate this reviewing.

## Tools

There are a lot of tools that can integrate with the PR review process to facilitate the manual review process for screenshots. A lot of these tools are geared for the web (e.g. [percy.io](https://percy.io/)), some are open source but not feature complete/usable for Zulip (e.g. [Pitsa](https://github.com/tvuotila/pitsa)).

I did come across a potentially very promising service: [Applitools](https://applitools.com/). Applitools notes that they support desktop apps and mobile. They also claim to have “advanced computer vision and AI” algorithms so they may be better to use than the aforementioned image comparison libraries.


## Conclusion

When I was developing the screenshot tests for windows/linux I was able to see how the app rendered on those platforms using their CIs and test suite. I found this to be much easier than having to setup a VM, clone the project, and run the app to verify things work as they should. A workflow that would have to be repeated for every platform.

But for visual regression testing we’ll need to use consistent machines for generating screenshots, perhaps CI machines only. Using CI machines as the source of truth for generating screenshots assumes that the CI services use consistent hardware, worst case scenario is we use privately hosted machines that we can control.

I’m curious about looking into a tool like [Applitools](https://applitools.com/) that can facilitate the PR review process and image comparison step. Not sure if would be more tooling/integration than is desired at this point in the project. This is also a paid service. Alternatively consider other free options. In the long term I’m not sure how maintainable it will be to commit images to the repo and try to use diffing algorithms with our own manually tuned thresholds. This maybe be a hindrance down the road.

## Some links from my research


* Collection of tools, talks, articles on Visual Regression testing: https://visualregressiontesting.com/
* Visual regression testing on Travis CI: http://twolfson.com/2014-02-25-visual-regression-testing-in-travis-ci
* [Video] Founder of Percy.io on their platform and the problem space of Visual Regression testing:  https://www.youtube.com/watch?v=iECnXK9EqA4&t=2611s
* Review of Visual vs. Functional Testing with pDiff and Applitools :https://saucelabs.com/blog/review-of-visual-vs-functional-testing-with-pdiff-and-applitools


@akashnimare @timabbott  thoughts ?

timabbott commented:
I think it might be reasonable to ask for free open source accounts for percy.io, applitools, and try out what they're doing (I think all these folks offer one) and see how much it helps?  My guess, supported by a lot of evidence from other startups that claim to use UI, is that it's fairly likely that the folks who claim to use AI are actually using humans for differences that are interesting, and planning to replace the humans with AI later.  I think using a third-party tool that's optimized for the image comparison stuff is probably a good plan, since it lets us focus on the other parts of making a good tool...

SimplyAhmazing commented:
I'll look into setting up Applitools next

akashnimare commented:
@SimplyAhmazing tests are still failing -
<details>
➜  zulip-electron git:(SimplyAhmazing-master) npm run test-e2e

> zulip@1.5.0 test-e2e /Users/akka/dev/zulipwork/zulip-electron
> gulp test-e2e

[18:17:01] Using gulpfile ~/dev/zulipwork/zulip-electron/gulpfile.js
[18:17:01] Starting 'test-e2e'...

TAP version 13
Detected App Data Dir base: /Users/akka/Library/Application Support
# app runs
PNG ERRRR null
ok 1 html title
not ok 2 screenshot comparison app-basic
  ---
    operator: ok
    expected: true
    actual:   false
    at: arePNGsLookSame (/Users/akka/dev/zulipwork/zulip-electron/node_modules/looks-same/index.js:205:72)

Saving screenshot, failed comparison: /Users/akka/dev/zulipwork/zulip-electron/tests/screenshots/darwin/app-basic-failed.png
  ...
Detected App Data Dir base: /Users/akka/Library/Application Support
# add-organization
PNG ERRRR null
ok 3 html title
not ok 4 screenshot comparison add-organization
  ---
    operator: ok
    expected: true
    actual:   false
    at: arePNGsLookSame (/Users/akka/dev/zulipwork/zulip-electron/node_modules/looks-same/index.js:205:72)

Saving screenshot, failed comparison: /Users/akka/dev/zulipwork/zulip-electron/tests/screenshots/darwin/add-organization-failed.png
  ...

1..4
# tests 4
# pass  2
# fail  2

[18:17:19] Finished 'test-e2e' after 18 s

npm ERR! Darwin 16.7.0
npm ERR! argv "/Users/akka/.nvm/versions/node/v6.9.4/bin/node" "/Users/akka/.nvm/versions/node/v6.9.4/bin/npm" "run" "test-e2e"
npm ERR! node v6.9.4
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE
npm ERR! zulip@1.5.0 test-e2e: `gulp test-e2e`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the zulip@1.5.0 test-e2e script 'gulp test-e2e'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the zulip package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     gulp test-e2e
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs zulip
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls zulip
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/akka/dev/zulipwork/zulip-electron/npm-debug.log
</details>

I'll test this on Windows as well.
